### ISSUE TITLE
为什么softmax之后求和不为1？

### ISSUE THREAD

**Issue by hhr114** [dmlc/dgl#3239]
## ❓ Questions and Help
我自己写了一个类似于 edge_softmax 的操作，因为我发现自带的 `dgl.nn.functional.edge_softmax` 不准，可是，我实现的版本也在最后有不小偏差，为什么？
```
import dgl.function as fn
z = nn.Parameter(torch.randn(size=(g.num_nodes(), 1))
g.ndata['w'] = torch.exp(z)
g.update_all(fn.copy_src(src='w', out='m'), reduce_func=fn.sum(msg='m', out='a')) 
g.apply_nodes(lambda nodes: {'att': nodes.data['w']/nodes.data['a']})
## to validate
g.update_all(fn.copy_src(src='att', out='m'), reduce_func=fn.sum(msg='m', out='p'))
```
Why is `g.ndata['p']` not an all-one vector?
![image.png](https://pic3.58cdn.com.cn/nowater/webim/big/n_v212196b9092024f1482d7f04fec01b9d6.png)

**Comment by yzh119** [comment]
I don't think your program acts like `edge_softmax`, edge softmax was applied to edges weight instead of node weights, but your program called a `apply_nodes` which was applied on node weights.

You should change your code to:
```python
import dgl.function as fn
z = nn.Parameter(torch.randn(size=(g.num_edges(), 1))
g.edata['w'] = torch.exp(z)
g.update_all(fn.copy_e(src='w', out='m'), reduce_func=fn.sum(msg='m', out='a')) 
g.apply_edges(lambda edges: {'att': edges.data['w']/edges.dst['a']})
## to validate
g.update_all(fn.copy_e(src='att', out='m'), reduce_func=fn.sum(msg='m', out='p'))
```
This should work.

**Comment by yzh119** [comment]
Also, could you please elaborate more on "因为我发现自带的 dgl.nn.functional.edge_softmax 不准"?
It would be great if you can provide specific test case.

**Comment by hhr114** [comment]
> I don't think your program acts like `edge_softmax`, edge softmax was applied to edges weight instead of node weights, but your program called a `apply_nodes` which was applied on node weights.
> 
> You should change your code to:
> 
> ```python
> import dgl.function as fn
> z = nn.Parameter(torch.randn(size=(g.num_edges(), 1))
> g.edata['w'] = torch.exp(z)
> g.update_all(fn.copy_e(src='w', out='m'), reduce_func=fn.sum(msg='m', out='a')) 
> g.apply_edges(lambda edges: {'att': edges.data['w']/edges.dst['a']})
> ## to validate
> g.update_all(fn.copy_e(src='att', out='m'), reduce_func=fn.sum(msg='m', out='p'))
> ```
> 
> This should work.

可是这样就违背了我计算的目的，物理意义是这样的，如果我有个节点数为N的图g，点i的置信度为实数$z_i$，所以z向量是N维实向量。我想用 softmax 算 点 i 在邻居中的相对置信度，即 $att_i=\frac{exp(z_i)}{\sum_{j\in N_i} exp(z_j)}$，请问除了我的代码写得有错吗？

**Comment by hhr114** [comment]
> Also, could you please elaborate more on "因为我发现自带的 dgl.nn.functional.edge_softmax 不准"?
> It would be great if you can provide specific test case.

我同时也试了以下代码，发现和手算的不一致
```
from dgl.nn.functional import edge_softmax
g.ndata['z'] = z
g.apply_edges(lambda edges: {'tmp': edges.src['z']})
g.edata['att'] = edge_softmax(g, g.edata.pop('tmp'))
```

**Comment by yzh119** [comment]
@hhr114 please set `norm_by` in `edge_softmax` to `'src'`, see https://docs.dgl.ai/generated/dgl.nn.functional.edge_softmax.html?highlight=edge_softmax#dgl.nn.functional.edge_softmax

**Comment by hhr114** [comment]
@yzh119 还是不一致，手算或者我之前写的ndata代码算出来第一个数是0.2897，我edge_softmax算出来是0.2996，用你给的代码算出来是0.2528。。哭了

**Comment by hhr114** [comment]
@yzh119 话说咱俩能加个微信实时交流一下吗？解决了我同样close the issue。。

**Comment by hhr114** [comment]
Thx
